docs/rfcs/component-status-reporting.md

```diff
@@ -0,0 +1,196 @@
+
+## The Goals the Component Health Reporting Should Achieve
+
+The following are the goals, as of June 2024 and with Collector 1.0 looming, for a component health reporting system.
+
+1. Runtime component health reporting, which cannot be automatically reported, must be opt-in for components.  Collector components must not be required to use the component health reporting system. This keeps component as compatible as possible with the Collector’s framework as we approach 1.0
+    - The consumers of the health reporting system must be able to identify which components are and are not using the health reporting system.
+2. Component health reporting must be opt-in for collector users.  While the underlying components are always allowed to report their health via the system, the Collector Host or any other listener may only take action when the user has configured the collector accordingly.
+   - As one example of compliance, the current health reporting system is dependent on the user configuring an extension that can watch for status updates.
+3. Component health must be representable as a finite state machine with clear transitions between states.
+4. The Collector Host may report statuses Starting, Ok,  Stopping and PermanentError on behalf of components as documented in the automation section.
+    - Additional status may be reported in the future
+5. Component health reporting must only be a mechanism for reporting health - it should have no mechanisms for taking actions on the health it reports.  How consumers of the health reporting system respond to component updates is not a concern of the health reporting system.
+
+## Existing deviations from those goals
+
+### Fatal Error Reporting
+
+Before the current implementation of component status reporting, a component could stop the collector by using `Host.ReportFatalError`. Now, a component MUST use component status reporting and emit a `FatalError`. This fact is in conflict with Goal 1, which states component health reporting must be opt-in for components.
+
+A couple solutions:
+1. Accept this reality as an exception to Goal 1.
+2. Add back `Host.ReportFatalError`.
+3. Remove the ability for components to stop the collector be removing `FatalError`.
```

Comment:
Sounds like the right future proof approach. We can and should have the collector decide of its behavior when an error is reported.

Reply:
If we remove this feature, we'd need to decide how components that are currently doing

```go
if errGrpc := jr.grpc.Serve(ln); !errors.Is(errGrpc, grpc.ErrServerStopped) && errGrpc != nil {
    jr.settings.ReportStatus(component.NewFatalErrorEvent(errGrpc))
}
```

Would handle the error returned from `.Serve`.

One option is for the component to return a `NewPermanent` error, but according to goal 1, component status reporting must be optional.  So for components that are not opt-in to component status reporting, and no longer have `Host.ReportFatalError`, what would they do?

Reply:
> Why isn't it acceptable to expect them report a permanent error?

At the moment I am working under the assumption that our current Status Reporting System is too new to be marked stable in component 1.0. Under that assumption, it needs to be ok for a component to NOT be using `ReportStatus` in order to interact with the Collector Host, which, until Status Reporting was added, was always done via `component.Host`.

If we want all the custom components out in the wild to be as compatible with component 1.0 as possible, then relying more on `component.Host` to interact with the Collector Host is better than status reporting.

Maybe there is a future where `component.Host` isn't needed anymore and components only use status reporting, but I don't think we're there yet. At the moment I view status reporting as a great feature that components SHOULD take advantage of, but are not required to.

Reply:
When you say "interact with the Collector Host", are we only talking about components being able to force the collector to shut down?

Reply:
Yes, but the other ways we interact with the host are accessing the available extensions and accessing available factories.

Reply:
It sounds like some notion of reporting a fatal error at any time is the only status-related interaction that we need to pin down for a 1.0 of component. I propose that we commit to such a mechanism for 1.0, but we _do not_ commit to a specific behavior in reaction to the event. 

This leaves us room to offer multiple strategies for reacting to a such an event, including:
- Immediate shutdown of the collector.
- Some automatic attempt to restart the component and/or component graph.
- Custom reactions directed via opamp.
- A full blown component status system which includes the notion of a fatal/permanent error.

Reply:
Thinking about this some more while refactoring component status features, I think the solution that provides the best separation of concerns is for the `component` package to provide no way for a component to stop the collector once the component has started. This is what the component's public API will represent once https://github.com/open-telemetry/opentelemetry-collector/pull/10725 is complete as we've already removed `Host.ReportFatalError`. 

With no public API in `component` to stop the collector after `Start`, the stance we'll be taking is: `Components cannot stop a running collector unless the `component.Host` implementation chooses to provide such a feature via a different Interface.`

This will end up being a breaking change for components that want to stop a running collector but do not want to use an unstable package like `componentstatus`.  Fortunately for us this breaking change was already handled when `Host.ReportFatalError` was removed, so any component that falls into this category is already in it.

Other components may choose to use `componentstatus.Reporter` and `componentstatus.NewFatalError`.  As part of the stabilization of `component` we do not need to commit to these interactions.